core/dagop.go

```diff
@@ -321,44 +339,384 @@ func (op RawDagOp) Exec(ctx context.Context, g bksession.Group, inputs []solver.
 	return []solver.Result{worker.NewWorkerRefResult(snap, opt.Worker)}, nil
 }
 
-func newDagOpLLB(ctx context.Context, dagOp buildkit.CustomOp, id *call.ID, inputs []llb.State) (llb.State, error) {
-	return buildkit.NewCustomLLB(ctx, dagOp, inputs,
-		llb.WithCustomNamef("%s %s", dagOp.Name(), id.Name()),
-		buildkit.WithTracePropagation(ctx),
-		buildkit.WithPassthrough(),
-	)
+func NewContainerDagOp(
+	ctx context.Context,
+	id *call.ID,
+	ctr *Container,
+	extraInputs []llb.State,
+) (*Container, error) {
+	mounts, inputs, err := getAllContainerMounts(ctr)
+	if err != nil {
+		return nil, err
+	}
+	inputs = append(inputs, extraInputs...)
+
+	dagop := &ContainerDagOp{
+		ID:     id,
+		Mounts: mounts,
+	}
+
+	st, err := newContainerDagOp(ctx, dagop, inputs)
+	if err != nil {
+		return nil, err
+	}
+
+	sts := make([]llb.State, 0, len(mounts))
+	for _, mount := range mounts {
+		if mount.Output == pb.SkipOutput {
+			sts = append(sts, llb.Scratch())
+		} else {
+			sts = append(sts, buildkit.StateIdx(ctx, st, mount.Output, nil))
+		}
+	}
+
+	ctr = ctr.Clone()
+	err = setAllContainerMounts(ctx, ctr, mounts, sts)
+	if err != nil {
+		return nil, err
+	}
+
+	return ctr, nil
 }
 
-type dagOpContextKey string
+func newContainerDagOp(
+	ctx context.Context,
+	dagop *ContainerDagOp,
+	inputs []llb.State,
+) (llb.State, error) {
+	if dagop.ID == nil {
+		return llb.State{}, fmt.Errorf("dagop ID is nil")
+	}
 
-func withDagOpContext(ctx context.Context, op buildkit.CustomOp) context.Context {
-	return context.WithValue(ctx, dagOpContextKey(op.Name()), op)
+	var t Container
+	requiredType := t.Type().NamedType
+	if dagop.ID.Type().NamedType() != requiredType {
+		return llb.State{}, fmt.Errorf("expected %s to be selected, instead got %s", requiredType, dagop.ID.Type().NamedType())
+	}
+
+	return newDagOpLLB(ctx, dagop, dagop.ID, inputs)
 }
 
-func DagOpFromContext[T buildkit.CustomOp](ctx context.Context) (t T, ok bool) {
-	if val := ctx.Value(dagOpContextKey(t.Name())); val != nil {
-		t, ok = val.(T)
+type ContainerDagOp struct {
+	ID *call.ID
+
+	Mounts []*pb.Mount
+
+	// Data is any additional data that should be passed to the dagop. It does
+	// not contribute to the cache key.
+	Data any
```

Comment:
what's the use case for this? didn't see it actually used in other dependent PRs, but might have missed it while sifting through